### ISSUE TITLE
no CNVs after run_numbat() and large run time

### ISSUE THREAD

**Issue by sbamopoulos** [kharchenkolab/numbat#98]
Hi Teng,

after finally being able to run the preprocessing script on all samples I am now encountering variable errors while executing run_numbat(). I would greatly appreciate your input on this. 

Some specifics of my data: 
- the datasets consist of bone marrow biopsies (all cell types) 
- the samples have a cell number of ~500-2000 cells and the input count matrix is a gene by cell_id sparse matrix (raw counts)
- as a reference I am using all cells of healthy samples from the same experiment (gene (~30000) by cell_id (~4500 cells) normalized gene count matrix)
- the allele dataframe is produced by the preprocessing script and looks like this (random lines): 
             cell        snp_id CHROM     POS        cM REF ALT AD DP  GT   gene
   Run_2_388793 1_1014228_G_A     1 1014228 0.7599337   G   A  1  1 1|0  ISG15
   Run_2_501211 1_1014228_G_A     1 1014228 0.7599337   G   A  1  1 1|0  ISG15
   Run_2_126885 1_1217251_C_A     1 1217251 1.1605514   C   A  1  1 1|0   SDF4
   Run_2_854627 1_1217251_C_A     1 1217251 1.1605514   C   A  0  1 1|0   SDF4
   Run_2_88065 1_1217251_C_A     1 1217251 1.1605514   C   A  0  1 1|0   SDF4
   Run_2_166838 1_1401954_G_T     1 1401954 1.2838815   G   T  0  1 0|1 MRPL20
   Run_2_468598 1_1401954_G_T     1 1401954 1.2838815   G   T  1  1 0|1 MRPL20
   Run_2_482968 1_1401954_G_T     1 1401954 1.2838815   G   T  1  1 0|1 MRPL20

1) The first issue that I encounter is that most samples seem to run indefinitely (exceeding the time limit of 4 days, and CPU/RAM limit of 32 cores and 64GB)

2) Furthermore, some samples produce the following error: 
INFO [2023-01-14 12:05:37] Running under parameters:
t = 0
alpha = 1e-04
gamma = 20
min_cells = 50
init_k = 3
max_cost = 198.6
max_iter = 2
max_nni = 100
min_depth = 0
use_loh = auto
multi_allelic = TRUE
min_LLR = 5
min_overlap = 0.45
max_entropy = 0
skip_nj = FALSE
diploid_chroms =
ncores = 32
ncores_nni = 32
common_diploid = TRUE
tau = 0.3
check_convergence = FALSE
plot = TRUE
genome = hg38
Input metrics:
662 cells
INFO [2023-01-14 12:05:37] Mem used: 8.29Gb
INFO [2023-01-14 12:06:33] Approximating initial clusters using smoothed expression ..
INFO [2023-01-14 12:06:34] Mem used: 8.29Gb
INFO [2023-01-14 12:06:41] running hclust...
INFO [2023-01-14 12:07:32] Iteration 1
INFO [2023-01-14 12:07:33] Mem used: 9.12Gb
INFO [2023-01-15 23:09:11] Running HMMs on 4 cell groups..
WARN [2023-01-15 23:09:17] No balanced segments, using all segments as baseline
INFO [2023-01-15 23:09:17] diploid regions: 1a,2a,3a,4a,5a,7a,8a,10a,11a,12a,15a,16a,17a,19a
INFO [2023-01-15 23:09:26] Expression noise level: high (1.4). Consider using a custom expression reference profile.
INFO [2023-01-15 23:09:27] No CNV remains after filtering by LLR in pseudobulks. Consider reducing min_LLR.

Setting min_LLR to 1 again results in a runtime exceeding ~4 days

3) Some samples procude the following error: 
INFO [2023-01-14 12:05:25] Running under parameters:
t = 0
alpha = 1e-04
gamma = 20
min_cells = 50
init_k = 3
max_cost = 429
max_iter = 2
max_nni = 100
min_depth = 0
use_loh = auto
multi_allelic = TRUE
min_LLR = 5
min_overlap = 0.45
max_entropy = 0
skip_nj = FALSE
diploid_chroms =
ncores = 32
ncores_nni = 32
common_diploid = TRUE
tau = 0.3
check_convergence = FALSE
plot = TRUE
genome = hg38
Input metrics:
1430 cells
INFO [2023-01-14 12:05:26] Mem used: 8.32Gb
INFO [2023-01-14 12:07:26] Approximating initial clusters using smoothed expression ..
INFO [2023-01-14 12:07:27] Mem used: 8.32Gb
INFO [2023-01-14 12:07:42] running hclust...
INFO [2023-01-14 12:09:26] Iteration 1
INFO [2023-01-14 12:09:27] Mem used: 10.1Gb
INFO [2023-01-16 05:56:16] Running HMMs on 4 cell groups..
WARN [2023-01-16 05:56:21] No balanced segments, using all segments as baseline
INFO [2023-01-16 05:56:21] diploid regions: 1a,2a,3a,4a,5a,6a,7a,9a,11a,12a,15a,17a,19a,20a
ERROR [2023-01-16 05:56:22] job 1 failed
ERROR [2023-01-16 05:56:22] Error in mutate(., state = run_joint_hmm(pAD = pAD, DP = DP, p_s = p_s,  :
  Problem while computing `state = run_joint_hmm(...)`.
ℹ The error occurred in group 8: CHROM = 9.
Caused by error in `h()`:
! error in evaluating the argument 'x' in selecting a method for function 'rowSums': (subscript) logical subscript too long

The produced output files until this error are:
exp_roll_clust.png  gexp_roll_wide.tsv.gz  hc.rds  log.txt  sc_refs.rds

Note that these samples have known large CNVs like e.g. del5q or trisomy.

1) Is there a possibility to reduce sample run time, without compromising the results too much, at least for testing?
2) What could be the reason for failure to detect CNVs (the recommendations for min_LLR and custom refmat do not seemt to work)?
3) I do not see any systematic to the last error, the samples should be fairly similar. Have you encountered this before?

I am looking forward to your input!

Best
Stefan

**Comment by teng-gao** [comment]
Hi Stefan,

Something is off because ~500-2000 cells shouldn't take more than 1 hour for Numbat to run. 

> as a reference I am using all cells of healthy samples from the same experiment (gene (~30000) by cell_id (~4500 cells) normalized gene count matrix)

How did you make the reference expression matrix? It should be gene by cell type, made from single-cell raw counts + cell type annotation using [`aggregate_counts`](https://kharchenkolab.github.io/numbat/reference/aggregate_counts.html).

Best,
Teng

**Comment by sbamopoulos** [comment]
Hi Teng,

you were right I just used a normalized expression matrix as a reference, which led to the large run times. The reference is still not annotated by cell type, so I used the default ref_hca as a reference for now. As you said it now runs within 1 hour, however it still does not produce any results, even after setting min_LLR to 1. Since I have a bone marrow sample ref_hca should cover most cell types. Is there any other setting/input that should be changed?

Best
Stefan
